.adr/0012-event-emission.md

```diff
@@ -0,0 +1,83 @@
+# 0012 Event Emission
+
+## Status
+
+Accepted
+
+## Context
+
+The existing mechanism for updates concerning objective updates is a go chan `ToApi()` on which the engine sends `EngineEvents`. The `go-nitro` Client spawns a go routine which reads those events, and then sends the objective identifier down either a `completedObjective` chan or `failedObjective`.
+
+The consuming application (h.f. consumer) would like to be able to synchronize its behaviour based on such updates. Currently, it is possible for the consumer to e.g. wait on a specific objective completing by reading from the `completedObjective` chan until the desired objective id is found. 
+
+The problem with this approach is that it is _not_ well suited for multiple consumers. If another module or goroutine in the consumer wants to wait on (say) a different objective, there is a likelihood that the first goroutine will read the identifier which the second goroutine is interested in.
+
+For example, let objective with `id1` complete and then objective with `id2` complete:
+
+```go
+func waitFor(objectiveId) {
+	for id := range client.completedObjectives {
+		if id == objectiveId {
+			return
+		}
+	}
+}
+
+waitFor(id2) // blocks until id1,id2 have been sent
+waitFor(id1) // blocks forever since id1 will not be re-sent
+
+```
+
+## Alternatives considered
+
+### Slice of `chans`
+
+It is possible to make a sort-of Javascript-style event system by having the `go-nitro` Client store a slice of `chans` for each objective id. Then, the consumer can "subscribe" by calling a method which appends a `chan` to that slice and returns it. Then, when the `client` learns from the engine that an objective has copmleted, it iterates through the slice and closes each `chan`. 
+
+This approach would not be idiomatic and has the following drawbacks:
+* if the subscriber is too late it can "miss" the event
+* trying to solve that issue with a cache leads to a race condition (a new subscription call reads the cache, infers the event has not fired, and then appends to the slice of `chans` _after_ the event fires and therefore still misses it)
+
+### Condition variable
+
+This is actually the most idiomatic approach if we want to emit events more than once. For example, if we wanted to have an "objective updated' event stream. See `sync.Cond` https://pkg.go.dev/sync#Cond .
```

Comment:
Instead of closing the channel couldn't we just leave the channel open and send empty structs, allowing us to implement "objective updated"?

One of the [links](https://blogtitle.github.io/go-advanced-concurrency-patterns-part-3-channels/) from the official documentation for `sync.Cond` says:

> [Cond](https://golang.org/pkg/sync/#Cond) is the most controversial type in the sync package. I think it’s a dangerous primitive and it is too easy to use incorrectly. I never use it because I don’t trust myself to use it right and during code review I always suggest to use other primitives instead. Even Bryan Mills (who is in the Go team and works a lot on sync primitives) got to the point of [proposing to remove it](https://github.com/golang/go/issues/21165).


which makes me slightly nervous 😟

Reply:
Interesting about `sync.Cond`. We could add this extra context to the ADR if you like. 

We can't simply send empty structs on the channel, because they would be read by only one consumer at a time. That doesn't give us a true "broadcast" pattern -- more like a "signal" pattern. 

We can also add this to the ADR to increase clarity.

Reply:
> We can't simply send empty structs on the channel, because they would be read by only one consumer at a time. That doesn't give us a true "broadcast" pattern -- more like a "signal" pattern.

Ah right!